packages/langium/src/lsp/completion/completion-provider.ts

```diff
@@ -158,19 +179,29 @@ export class DefaultCompletionProvider implements CompletionProvider {
         await Promise.all(
             stream(beforeFeatures)
                 .distinct(distinctionFunction)
-                .map(e => this.completionFor(context, e, acceptor))
+                .map(e => this.completionForContexts(contexts, e, acceptor))
         );
 
         if (reparse) {
             await Promise.all(
                 stream(afterFeatures)
                     .exclude(beforeFeatures, distinctionFunction)
                     .distinct(distinctionFunction)
-                    .map(e => this.completionFor(context, e, acceptor))
+                    .map(e => this.completionForContexts(contexts, e, acceptor))
             );
         }
 
-        return CompletionList.create(items, true);
+        return CompletionList.create(this.deduplicateItems(items), true);
```

Comment:
Is there an explanation for why exactly duplicate items can appear? Can we prevent that at the root cause instead of post-processing?

Reply:
There are a lot of edge cases in which the scope provider might return an element multiple times during the course of the completion. For example:

```
Rule: ref1=[Definition:ID] 'x' | ref2=[Definition:ID] 'y';
```

Or

```
Rule: ref1=[Super:ID] 'x' | ref2=[Sub:ID] 'y';

interface Sub extends Super { }
```

In both cases we have different cross references which yield some of the same elements. We do try to prevent duplicates already beforehand (see [here](https://github.com/langium/langium/blob/c2797380edc01ed170d0f9ada926551209b01fcf/packages/langium/src/lsp/completion/completion-provider.ts#L160) and [here](https://github.com/langium/langium/blob/c2797380edc01ed170d0f9ada926551209b01fcf/packages/langium/src/lsp/completion/completion-provider.ts#L274)), but some of these are not preventable through these measures.